edb/lang/edgeql/parser/grammar/expressions.py

```diff
@@ -314,11 +314,12 @@ def reduce_ShapeElementWithSubShape(self, *kids):
 
         shape = kids[1].val
 
-        if shape and isinstance(shape[0], qlast.ObjectRef):
+        # shape elements can have a path starting with a TypeName,
+        # this indicates a polymorphic shape and TypeName must be
+        # extracted from the path steps
+        if shape and isinstance(shape[0], qlast.TypeName):
```

Comment:
Polymorphic shapes should support `FullTypeExpr`

Reply:
OK, changing to `TypeExpr`.